binr/ragg2/ragg2.c

```diff
@@ -108,6 +123,454 @@ static int openfile (const char *f, int x) {
+	  			cEnv->CC, cEnv->CFLAGS, file, file);
+	if (output == NULL) {
+		eprintf ("Compilation failed!\n");
+		goto fail;
+	}
+	printf ("%s", output);
+
+	if ((fileExt = r_str_newf ("%s.s", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+
+	fileExtFd = open (fileExt, O_CREAT | O_RDWR, 0644);
+	if (fileExtFd == -1) {
+		eprintf ("Error while opening %s.s\n", file);
+		goto fail;
+	}
+	write (fileExtFd, cEnv->SHDR, strlen (cEnv->SHDR));
+	close (fileExtFd);
+
+	free (output);
+	output = r_sys_cmd_strf ("cat '%s.tmp' | sed -e s,rdata,text, -e 's,get_pc_thunk.bx,__getesp__,g' "
+	  			 "| grep -v .cstring | grep -v size | grep -v ___main | grep -v section "
+				 "| grep -v __alloca | grep -v zero | grep -v cfi >> '%s.s'", file, file);
+	if (!output) {
+		eprintf ("Write in %s.s failed!\n", file);
+		goto fail;
+	}
+	printf ("%s", output);
+
+	printf ("==> Assemble\n");
+	printf ("%s %s -Os -o '%s.o' '%s.s'\n", cEnv->CC, cEnv->LDFLAGS, file, file);
+
+	free (output);
+	output = r_sys_cmd_strf ("%s %s -Os -o '%s.o' '%s.s'", cEnv->CC, cEnv->LDFLAGS, file, file);
+	if (!output) {
+		eprintf ("Assembly failed!\n");
+		goto fail;
+	}
+	printf ("%s", output);
+
+	printf ("==> Link\n");
+	printf ("rabin2 -o '%s.text' -O d/S/%s '%s.o'\n", file, cEnv->TEXT, file);
+
+	free (output);
+	output = r_sys_cmd_strf ("rabin2 -o '%s.text' -O d/S/%s '%s'.o", file, cEnv->TEXT, file);
+	if (!output) {
+		eprintf ("Linkage failed!\n");
+		goto fail;
+	}
+
+	free (fileExt);
+	if ((fileExt = r_str_newf ("%s.o", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+
+	if (!r_file_exists (fileExt)) {
+		eprintf ("Cannot find %s.o\n", file);
+		goto fail;
+	}
+
+	free (fileExt);
+	if ((fileExt = r_str_newf ("%s.text", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+	if (r_file_size (fileExt) == 0) {
+		printf ("FALLBACK: Using objcopy instead of rabin2");
+
+		free (output);
+		output = r_sys_cmd_strf ("%s -j .text -O binary '%s.o' '%s.text'", 
+		  		cEnv->OBJCOPY, file, file);
+		if (!output) {
+			eprintf ("objcopy failed!\n");
+			goto fail;
+		}
+	}
+
+	free (fileExt);
+	if ((fileExt = r_str_newf ("%s.bin", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+	r_file_rm (fileExt);
+
+	free (fileExt);
+	if ((fileExt = r_str_newf ("%s.tmp", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+	r_file_rm (fileExt);
+
+	free (fileExt);
+	if ((fileExt = r_str_newf ("%s.s", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+	r_file_rm (fileExt);
+
+	free (fileExt);
+	if ((fileExt = r_str_newf ("%s.o", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+	r_file_rm (fileExt);
+
+	free (fileExt);
+	if ((fileExt = r_str_newf ("%s.text", file)) == NULL) {
+		eprintf ("Error with r_str_newf!\n");
+		goto fail;
+	}
+
+	free (output);
+	free_cEnv (cEnv);
```

Comment:
Spaguetti

Reply:
By spaguetti you are talking about the removal of the extended files (.o .s .tmp ...) :) ? I will use an array then.